### PR TITLE
Refactor context-value patten to hide context-key from consumers

### PR DESCRIPTION
**Pull request by vhvb1989**
The context-key-value feature enabled a go-context to add any value to a context using a `key`.
Then, the `value` can be fetch from the context gerarchy.

The go-context is implemented as a simple linked list, so every time a value is added to a context,
a new node is added to the linked list. When looking for some key-value, the linked list is traversed
back while trying to find a key.

When any component uses this feature and expects to fetch some key-value from a context,
we should follow a pattern which avoids coupling the component to any other component 
that sets the key-value.
For example, if component `A` will fetch the key `key-a` from the context, and component `B` will
set a value for `key-a`, component `B` *must not* depend on knowing what's the name of the key to 
use for the value, since that would create a hard dependency between A and B.
Instead, component `A` should expose an API that produces the context, like:
`func WithKey(ctx context.Context, value ValueType) context.Context`
This API is consumed by component `B`, which doesn't really care about the key used for the value.
If `A` or any other component need to get the value, there should be another function like:
`func ValueTypeFromContext(ctx context.Context) ValueType`
This function is used whenever the value wants to be retrieved.

This PR updates the implementation from azd to follow the context-value pattern as described below
and breaks the current dependencies between components (for example, everyone depending on `environment`
package just to know the keys for settings the context)

- hide optionsContextKey behind commands package
- hide azdContextKey behind azdContext
- move azd context to be a package and not part of environment package
- hide contextKey behind azcli package
- remove AzdCliContextKey
- remove TemplateContextKey
- remove HttpUtilContextKey from environment

### CODE REVIEW

File: cli/azd/pkg/commands/builder.go [hunk 1]
```diff
@@ -16,18 +17,18 @@ func Build(action Action, rootOptions *GlobalCommandOptions, use string, short s
...
-			azdCtx, err := environment.NewAzdContext()
+			azdCtx, err := azd_context.NewAzdContext()
 			if err != nil {
 				return fmt.Errorf("creating context: %w", err)
 			}
 
 			// Set the global options in the go context
-			ctx = context.WithValue(ctx, environment.AzdContextKey, azdCtx)
-			ctx = context.WithValue(ctx, environment.OptionsContextKey, rootOptions)
+			ctx = azd_context.WithAzdContext(ctx, azdCtx)
```
Comment (wbreza) [inline cli/azd/pkg/commands/builder.go:26]: If we're going to store the `AzdContext` in the context, then I think we should avoid passing around the `azdCtx` variable so much through the rest of the system.

**Review by wbreza — COMMENTED** [review]
Overall I really like this and was just reading a similar article about pitfalls of go context and recommending this style of approach.

If we are going to do this now I also recommend we get rid of passing around `AzdContext` and instead just get it when we need it.

File: cli/azd/pkg/commands/builder.go (hunk 1 above)
Comment (vhvb1989) [reply cli/azd/pkg/commands/builder.go:26]: Yes, I can create a subsequent PR for that.

I would like us to start creating patterns and coding guidelines around our components.

**Comment by ellismg** [comment]
> move azd context to be a package and not part of environment package

If we are going to do this (I am guessing the thing that drove this is just separating these since their concerns are slightly different) could we name the new package `azdcontext`?  Having the underscore in the package name isn't the best ergonomics, IMHO.

**Comment by vhvb1989** [comment]
> > move azd context to be a package and not part of environment package
> 
> If we are going to do this (I am guessing the thing that drove this is just separating these since their concerns are slightly different) could we name the new package `azdcontext`? Having the underscore in the package name isn't the best ergonomics, IMHO.

Yes, because some components uses/depend-on azdcontext, but doesn't require anything from environment.

I will rename to `azdcontext`

**Comment by vhvb1989** [comment]
> > move azd context to be a package and not part of environment package
> 
> If we are going to do this (I am guessing the thing that drove this is just separating these since their concerns are slightly different) could we name the new package `azdcontext`? Having the underscore in the package name isn't the best ergonomics, IMHO.

Also because the `azdcontext` will now expose the API to add AzdContext value to the context.  Otherwise, the `environment` package would need to provide that and it would read weird as `environment.WithAzdContext(...)`  vs `azdcontext.WithAzdContenxt(...)`

**Review by wbreza — APPROVED** [review]

**Review by ellismg — APPROVED** [review]
LGTM

**Pull request merged**